modules/cluster_estimation/cluster_estimation_by_label.py

```diff
@@ -0,0 +1,162 @@
+
+        # Cluster model corresponding to each label
+        # Each cluster estimation object stores the detections given to in its __all_points bucket across runs
+        self.__label_to_cluster_estimation_model: dict[
+            int, cluster_estimation.ClusterEstimation
+        ] = {}
+
+    def run(
+        self,
+        input_detections: list[detection_in_world.DetectionInWorld],
+        run_override: bool,
+    ) -> tuple[True, dict[int, list[object_in_world.ObjectInWorld]]] | tuple[False, None]:
+        """
+        See `ClusterEstimation` for parameter descriptions.
+
+        RETURNS
+        -------
+        model_ran: bool
+            True if ClusterEstimation object successfully ran its estimation model, False otherwise.
+
+        labels_to_objects: dict[int, list[object_in_world.ObjectInWorld] or None.
+            Dictionary where the key is a label and the value is a list of all cluster detections with that label.
+            ObjectInWorld objects don't have a label property, but they are sorted into label categories in the dictionary.
+        """
+        label_to_detections: dict[int, list[detection_in_world.DetectionInWorld]] = {}
+
+        # Filtering detections by label
+        for detection in input_detections:
+            if not detection.label in label_to_detections:
+                label_to_detections[detection.label] = []
+
+            label_to_detections[detection.label].append(detection)
+
+        labels_to_objects: dict[int, list[object_in_world.ObjectInWorld]] = {}
+
+        for label, detections in label_to_detections.items():
+            # Create cluster estimation for label if it doesn't exist
+            if not label in self.__label_to_cluster_estimation_model:
+                result, cluster_model = cluster_estimation.ClusterEstimation.create(
+                    self.__min_activation_threshold,
+                    self.__min_new_points_to_run,
+                    self.__max_num_components,
+                    self.__random_state,
+                    self.__local_logger,
+                )
+                if not result:
+                    self.__local_logger.error(
+                        f"Failed to create cluster estimation for label {label}"
+                    )
+                    return False, None
+
+                self.__label_to_cluster_estimation_model[label] = cluster_model
+
+            # Runs cluster estimation for specific label
+            result, clusters = self.__label_to_cluster_estimation_model[label].run(
+                detections,
+                run_override,
+            )
+
+            if not result:
+                self.__local_logger.error(
+                    f"Failed to run cluster estimation model for label {label}"
+                )
+                return False, None
+
+            if not label in labels_to_objects:
+                labels_to_objects[label] = []
+            labels_to_objects[label] += clusters
```

Comment:
Is this the desired behaviour? The cluster estimation objects already hold a record of all points, so they will always generate an updated version of the cluster centres. I think this should be an unconditional assignment instead.